golang/cosmos/x/vstorage/keeper/keeper.go

```diff
@@ -129,7 +141,31 @@ func (k Keeper) ImportStorage(ctx sdk.Context, entries []*types.DataEntry) {
 	for _, entry := range entries {
 		// This set does the bookkeeping for us in case the entries aren't a
 		// complete tree.
-		k.SetStorage(ctx, entry.Path, entry.Value)
+		k.SetStorage(ctx, types.NewStorageEntry(entry.Path, entry.Value))
+	}
+}
+
+func (k Keeper) MigrateNoDataPlaceholders(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+
+	// Copy empty keys first since cosmos stores do not support writing keys
+	// while an iterator is open over the domain
+	emptyKeys := []*([]byte){}
+	{
+		defer iterator.Close()
+		for ; iterator.Valid(); iterator.Next() {
+			rawValue := iterator.Value()
+			if len(rawValue) == 1 && rawValue[0] == types.EncodedDataPrefix[0] {
```

Comment:
This isn't wrong, but I'd feel more comfortable with an equivalent that compares fully.
```suggestion
			if string(rawValue) == string(types.EncodedDataPrefix) {
```

Reply:
Fine by me. Is there a runtime cost to these conversions ?

Reply:
Ah yes I remember why I did this, it was definitely a perf concern not knowing the cost of byte to string "conversion", in case the `rawValue` is a lot more than a single char.

Reply:
The compiler _should_ take care of that because the internal [`bytestring`](https://go.dev/ref/spec#Core_types) representations are identical, but you'd need `go bench` or assembly inspection to verify that it does in fact optimize away the [mutable ↔︎ immutable conversion](https://go.dev/ref/spec#Conversions_to_and_from_a_string_type). Regardless, there's a reasonable guaranteed-bounded approach:
```suggestion
			if len(rawValue) == len(types.EncodedDataPrefix) &&
				string(rawValue) == string(types.EncodedDataPrefix) {
```

Reply:
Ah, I've just been out of Go for too long. There is in fact an actually correct way to do this: https://pkg.go.dev/bytes#Equal
```suggestion
			if bytes.Equal(rawValue, types.EncodedDataPrefix) {
```

Reply:
Duh!

Reply:
> cannot use types.EncodedDataPrefix[0] (variable of type byte) as []byte value in argument to bytes.Equal
> compiler [IncompatibleAssign](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#IncompatibleAssign)